### PR TITLE
Mixological Masterful: Makes shaking drinks a thing

### PR DESCRIPTION
**Pull request by EOBGames**
## About The Pull Request
Changes up shakers to be shakeable, and adds a shaking requirement to a series of drinks.

Alt-click caps and uncaps the shaker. While it's sealed, clicking inhand shakes. While uncapped, stuff can be poured out or in.

Shaken drinks:
- Vodka Martini
- Pan-Galactic Gargle Blaster
- Toxins Special
- Beepsky Smash
- Irish Cream
- Margarita
- Bahama Mama
- Antifreeze
- Painkiller
- Changeling Sting
- Devil's Kiss
- Banana Honk
- Silencer
- Thirteen Loko
- Whiskey Sour
- Eggnog
- Quadruple Sec
- Grasshopper
- Quintuple Sec
- Alexander
- Sidecar
- Between the Sheets
- Kamikaze
- Branca Menta
- Jack Rose
- Blazaam
- Mauna Loa
- Amaretto Alexander
- Yuyakita
- Banzai Ti
- Imperial Mai Tai
- Konococo Rumtini
- Blue Hawaiian
- Maui Sunrise
- Protein Blend
- Sea Breeze
- Salt and Swell
- Tich Toch
- Tropical Storm

Drinks were chosen based on a few factors- any real life drink that is shaken was added, and for fictional drinks, I generally picked anything that had an emulsifier (egg white or blood, mostly), pineapple juice or citrus juice since these are general telltales for shaking.

Big thanks to Melbert for providing the shake code; without it, this PR would really not exist.
## Why It's Good For The Game
Gives a little more tactile authenticity to the bartender.
## Changelog
:cl: EOBGames, MrMelbert
add: Shake to your heart's content with the new working shakers! Certain drinks will now need shaken before they're ready to enjoy; for a full list, see the PR.
/:cl:

### CODE REVIEW

**Comment by EricZilla** [comment]
How satisfying is the shake sound, is what we all want to know (probably just me)

**Comment by CRITAWAKETS** [comment]
I'd prefer my drinks _stirred_, not shaken, thank you.

**Comment by SomeoneYouProbablyKnow** [comment]
Is the shaker the only implement you can use to shake drinks? Can you circumvent it in some way? But really locking the funny sec heal drink behind the shaker is pretty neat, no more mute sec robbing bar.

**Comment by AdipemDragon** [comment]
could the reaction search function also say if something needs shaking, just so there's an ingame way to check besides just making your drink and watching it not mix

**Comment by private-tristan** [comment]
does this mean that bartender is effectively unable to make half their drinks if their shaker gets stolen roundstart/before they join?

**Comment by MrMelbert** [comment]
> does this mean that bartender is effectively unable to make half their drinks if their shaker gets stolen roundstart/before they join?

you can print shakers in the lathe

**Comment by san7890** [comment]
add gravy i've been wanting people to be able to mix gravy in a bottle for at least 14 months now

image

**Comment by Justice12354** [comment]
> add gravy i've been wanting people to be able to mix gravy in a bottle for at least 14 months now
> 
> (image)
> 
> this (was) on the the official blender bottle site btw https://web.archive.org/web/20220629005500/https://www.blenderbottle.com/blogs/recipes/never-fail-blenderbottle-gravy

Make it so when people shake gravy, they say "Oh, so smooth!"

**Comment by Loafin34** [comment]
i need the gravy drink. be sure to give it a special interaction with being splashed on people by letting other people eat them for 5 brute damage (because the gravy makes them smell so good)

**Comment by Wallemations** [comment]
I thought the entire lid's supposed to come off, but only the straining cap gets opened at any time
It's really just up to you, but imo if you removed the whole lid it would make it more visibly obvious which state it's in?

**Comment by Livrah** [comment]
Can you still make those cocktails with service-chem plumbing?

**Comment by EOBGames** [comment]
> I thought the entire lid's supposed to come off, but only the straining cap gets opened at any time It's really just up to you, but imo if you removed the whole lid it would make it more visibly obvious which state it's in?

On a real cobbler, yeah, it would ordinarily open up fully- that said, that's also more shading and really general sprite quality to worry about, which is why I didn't do it.

**Comment by newfren** [comment]
it would be nice if there was a way to shake these without relying solely on the shaker - maybe the mix setting of a grinder?

**Comment by OmegaDarkPotato** [comment]
Just to be certain- service borgs are still able to produce these drinks with their shaker, yes?

**Comment by carshalash** [comment]
Controls seem pretty reasonable and accessable to handle. Doesn't interfere with normal drink transference rates which is good.

Only recipe i'm questionable about is changeling sting, isn't that one made with a soda?

File: code/modules/reagents/reagent_containers/cups/drinks.dm [hunk 1]
```diff
@@ -434,13 +434,92 @@
...
+	. |= AFTERATTACK_PROCESSED_ITEM
+
+	if(capped && (target.is_refillable() || target.is_drainable() || (reagents.total_volume && !user.combat_mode)))
+		to_chat(user, span_warning("You must remove the cap before you can do that!"))
+		return
+
+	return . | ..()
+
+/obj/item/reagent_containers/cup/glass/shaker/attack_self(mob/user)
+	if(!is_drainable())
```
Comment (Jacquerel) [inline code/modules/reagents/reagent_containers/cups/drinks.dm:511]:
```suggestion
	if(is_drainable())
		return ..()
```

File: code/modules/reagents/reagent_containers/cups/drinks.dm [hunk 2]
```diff
@@ -434,13 +434,92 @@
...
+	if(capped)
+		return FALSE
+	return ..()
+
+/obj/item/reagent_containers/cup/glass/shaker/attack(mob/target, mob/living/user, def_zone)
+	if(!target)
+		return
+
+	if(capped && reagents.total_volume && istype(target))
+		to_chat(user, span_warning("You must remove the cap before you can do that!"))
```
Comment (Jacquerel) [inline code/modules/reagents/reagent_containers/cups/drinks.dm:496]:
```suggestion
		balloon_alert(user, "remove the cap!")
```

File: code/modules/reagents/reagent_containers/cups/drinks.dm [hunk 3]
```diff
@@ -434,13 +434,92 @@
...
+		return
+
+	return ..()
+
+/obj/item/reagent_containers/cup/glass/shaker/afterattack(obj/target, mob/living/user, proximity)
+	. |= AFTERATTACK_PROCESSED_ITEM
+
+	if(capped && (target.is_refillable() || target.is_drainable() || (reagents.total_volume && !user.combat_mode)))
+		to_chat(user, span_warning("You must remove the cap before you can do that!"))
```
Comment (Jacquerel) [inline code/modules/reagents/reagent_containers/cups/drinks.dm:505]:
```suggestion
		balloon_alert(user, "remove the cap!")
```

File: code/modules/reagents/reagent_containers/cups/drinks.dm [hunk 4]
```diff
@@ -434,13 +434,92 @@
...
+/obj/item/reagent_containers/cup/glass/shaker/attack_self(mob/user)
+	if(!is_drainable())
+		balloon_alert(user, "shaking...")
+		Shake(duration = 1 SECONDS)
+		if(reagents.total_volume)
+			playsound(loc, 'sound/effects/can_shake.ogg', 50, TRUE)
+		if(!do_after(user, 1 SECONDS, target = user))
+			return FALSE
+		shaken = TRUE
+		addtimer(VARSET_CALLBACK(src, shaken, FALSE), 2 SECONDS)
```
Comment (Jacquerel) [inline code/modules/reagents/reagent_containers/cups/drinks.dm:519]:
```suggestion
		addtimer(VARSET_CALLBACK(src, shaken, FALSE), 2 SECONDS, TIMER_DELETE_ME)
```
Actually don't know if this is required but might stop something funny happening if you delete the shaker in the 2 seconds before the timer fires.

File: code/modules/reagents/reagent_containers/cups/drinks.dm [hunk 5]
```diff
@@ -434,13 +434,92 @@
...
+	else
+		. += span_notice("The cap has been taken off and the shaker can be filled or emptied. Alt-click to put a cap on.")
+
+/obj/item/reagent_containers/cup/glass/shaker/AltClick(mob/user)
+	. = ..()
+	if(capped)
+		capped = FALSE
+		spillable = TRUE
+		animate(src, transform = null, time = 2, loop = 0)
+		balloon_alert(user, "cap removed...")
```
Comment (Jacquerel) [inline code/modules/reagents/reagent_containers/cups/drinks.dm:472]:
```suggestion
		balloon_alert(user, "cap removed")
```
Ellipses are only for progress bars.

File: code/modules/reagents/reagent_containers/cups/drinks.dm [hunk 6]
```diff
@@ -434,13 +434,92 @@
...
+		playsound(loc, 'sound/effects/can_open1.ogg', 50, TRUE)
+	else
+		capped = TRUE
+		spillable = FALSE
+		balloon_alert(user, "cap added...")
```
Comment (Jacquerel) [inline code/modules/reagents/reagent_containers/cups/drinks.dm:477]:
```suggestion
		balloon_alert(user, "cap added")
```

File: code/modules/reagents/reagent_containers/cups/drinks.dm [hunk 7]
```diff
@@ -434,13 +434,92 @@
 	amount_per_transfer_from_this = 10
 	volume = 100
 	isGlass = FALSE
+	var/capped = TRUE
+	var/shaken = FALSE
+	var/cap_icon = 'icons/obj/drinks/bottles.dmi'
+	var/cap_icon_state = "shaker_cap"
+	var/mutable_appearance/cap_overlay
```
Comment (Jacquerel) [inline code/modules/reagents/reagent_containers/cups/drinks.dm:441]: Document these vars

**Review by Jacquerel — CHANGES_REQUESTED** [review]
> could the reaction search function also say if something needs shaking, just so there's an ingame way to check besides just making your drink and watching it not mix

I think this would be a pretty valuable thing to add before the PR is mergeable, although IDK what form it would take.

**Comment by randy10122** [comment]
genuinely considering learning how to work with dm just to finish up this pr. this is too cool to be forgotten

**Comment by san7890** [comment]
Good luck

**Pull request closed without merging**